### PR TITLE
[Feature][Connector-V2] Add Huawei Cloud OBS connector

### PR DESCRIPTION
**Pull request by kim-up**
#4577

## Check list

* [x] Code changed are covered with tests, or it does not need tests for reason:
* [x] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [x] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs
* [x] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/incubator-seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/incubator-seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-dist/pom.xml)
* [x] Update the [`release-note`](https://github.com/apache/incubator-seatunnel/blob/dev/release-note.md).

### CODE REVIEW

**Comment by TyrantLucifer** [comment]
Good pull request, let's waiting CI

**Comment by kim-up** [comment]
image

The dependency of `esdk-obs-java` can not download. 
https://mvnrepository.com/artifact/com.huawei.storage/esdk-obs-java/3.19.7.3
image

@TyrantLucifer How should this be handled?

**Comment by TyrantLucifer** [comment]
> image
> 
> The dependency of `esdk-obs-java` can not download. https://mvnrepository.com/artifact/com.huawei.storage/esdk-obs-java/3.19.7.3 image
> 
> @TyrantLucifer How should this be handled?

You can find solution in `oss-jindo` module's pom file.

File: docs/en/connector-v2/sink/ObsFile.md [hunk 1]
```diff
@@ -0,0 +1,260 @@
+# ObsFile
```
Comment (EricJoy2048) [inline docs/en/connector-v2/sink/ObsFile.md:1]: After discussion in the community, we have optimized the format requirements of the document, which you can refer to：`https://github.com/apache/incubator-seatunnel/issues/4544`

Comment (kim-up) [reply docs/en/connector-v2/sink/ObsFile.md:1]: Ok

**Comment by hailin0** [comment]
Please add e2e testcase and make junit @Disabled

**Comment by kim-up** [comment]
@TyrantLucifer @EricJoy2048 @hailin0  Is there anything else I need to do ?

**Comment by kim-up** [comment]
@TyrantLucifer @hailin0 @EricJoy2048 @jbonofre  Asking for Help!. Are there any end-to-end test examples for file storage cloud services such as obs, oss, and s3? Because it involves sensitive information such as ak and sk. And cloud services do not necessarily have relevant docker images that can be used for Mock testing.

**Comment by TyrantLucifer** [comment]
> 

Hi @kim-up , you can add a e2e test case and run it in your local env, before submit codes you should add annotation to disabled it and remove sensitive information from config file. Finally offer the snapshot image in this pull request. You can refer to `selectdb-e2e` and `file-local-e2e`. The idea is to allow novices to quickly find test cases for the connector and run through the code framework you provide by directly changing some key information

**Comment by kim-up** [comment]
> > 
> 
> Hi @kim-up , you can add a e2e test case and run it in your local env, before submit codes you should add annotation to disabled it and remove sensitive information from config file. Finally offer the snapshot image in this pull request. You can refer to `selectdb-e2e` and `file-local-e2e`. The idea is to allow novices to quickly find test cases for the connector and run through the code framework you provide by directly changing some key information

Get it.

**Comment by kim-up** [comment]
Docs & e2e have  add. Start  the workflow? @TyrantLucifer @EricJoy2048 @hailin0

**Comment by TyrantLucifer** [comment]
> Docs & e2e have add. Start the workflow? @TyrantLucifer @EricJoy2048 @hailin0

Approved, let's waiting CI/CD, BTW could you please offer some snapshots of this connector?

**Comment by kim-up** [comment]
> > Docs & e2e have add. Start the workflow? @TyrantLucifer @EricJoy2048 @hailin0
> 
> Approved, let's waiting CI/CD, BTW could you please offer some snapshots of this connector?

What is  `some snapshots` ? Docker image ?  Is there any related document that can refer to...

**Comment by kim-up** [comment]
## Local Test Images
### Csv Test
#### fake_to_obs_csv
```
env {
  execution.parallelism = 1
  checkpoint.interval = 5000
  job.mode = "BATCH"
}

source {
  FakeSource {
    result_table_name = "fake"
    schema = {
      fields {
        c_map = "map<string, string>"
        c_array = "array<int>"
        c_string = string
        c_boolean = boolean
        c_tinyint = tinyint
        c_smallint = smallint
        c_int = int
        c_bigint = bigint
        c_float = float
        c_double = double
        c_bytes = bytes
        c_date = date
        c_decimal = "decimal(38, 18)"
        c_timestamp = timestamp
        c_row = {
          c_map = "map<string, string>"
          c_array = "array<int>"
          c_string = string
          c_boolean = boolean
          c_tinyint = tinyint
          c_smallint = smallint
          c_int = int
          c_bigint = bigint
          c_float = float
          c_double = double
          c_bytes = bytes
          c_date = date
          c_decimal = "decimal(38, 18)"
          c_timestamp = timestamp
        }
      }
    }
  }
}

sink {
  ObsFile {
    path="/seatunnel/csv"
    bucket = "obs://********"
    access_key = "********"
    access_secret = "********"
    endpoint = "obs.cn-east-3.myhuaweicloud.com"
    partition_dir_expression="${k0}=${v0}"
    is_partition_field_write_in_file=true
    file_name_expression="${transactionId}_${now}"
    file_format_type="csv"
    filename_time_format="yyyy.MM.dd"
    is_enable_transaction=true
  }
}
```
##### Result
image
image


#### obs_csv_projection_to_assert
```
env {
  execution.parallelism = 1
  checkpoint.interval = 5000
  job.mode = "BATCH"
}

source {
  ObsFile {
    path="/seatunnel/csv"
    bucket = "obs://******"
    access_key = "******"
    access_secret = "******"
    endpoint = "obs.cn-east-3.myhuaweicloud.com"
    result_table_name = "fake"
    file_format_type = csv
    delimiter = ","
    read_columns = [c_string, c_boolean]
    skip_header_row_number = 1
    schema = {
      fields {
          c_map = "map<string, string>"
          c_array = "array<int>"
          c_string = string
          c_boolean = boolean
          c_tinyint = tinyint
          c_smallint = smallint
          c_int = int
          c_bigint = bigint
          c_float = float
          c_double = double
          c_bytes = bytes
          c_date = date
          c_decimal = "decimal(38, 18)"
          c_timestamp = timestamp
          c_row = {
            c_map = "map<string, string>"
            c_array = "array<int>"
            c_string = string
            c_boolean = boolean
            c_tinyint = tinyint
            c_smallint = smallint
            c_int = int
            c_bigint = bigint
            c_float = float
            c_double = double
            c_bytes = bytes
            c_date = date
            c_decimal = "decimal(38, 18)"
            c_timestamp = timestamp
          }
      }
    }
  }
}


sink {
  Assert {
    rules {
      row_rules = [
        {
          rule_type = MAX_ROW
          rule_value = 5
        }
      ],
      field_rules = [
        {
          field_name = c_string
          field_type = string
          field_value = [
            {
              rule_type = NOT_NULL
            }
          ]
        },
        {
          field_name = c_boolean
          field_type = boolean
          field_value = [
            {
              rule_type = NOT_NULL
            }
          ]
        }
      ]
    }
  }
}
```
##### Result
image

**Comment by kim-up** [comment]
## Local Test Images
### Excel Test
#### fake_to_obs_excel
```
env {
  execution.parallelism = 1
  checkpoint.interval = 5000
  job.mode = "BATCH"
}

source {
  FakeSource {
    result_table_name = "fake"
    schema = {
      fields {
        c_map = "map<string, string>"
        c_array = "array<int>"
        c_string = string
        c_boolean = boolean
        c_tinyint = tinyint
        c_smallint = smallint
        c_int = int
        c_bigint = bigint
        c_float = float
        c_double = double
        c_bytes = bytes
        c_date = date
        c_decimal = "decimal(38, 18)"
        c_timestamp = timestamp
        c_row = {
          c_map = "map<string, string>"
          c_array = "array<int>"
          c_string = string
          c_boolean = boolean
          c_tinyint = tinyint
          c_smallint = smallint
          c_int = int
          c_bigint = bigint
          c_float = float
          c_double = double
          c_bytes = bytes
          c_date = date
          c_decimal = "decimal(38, 18)"
          c_timestamp = timestamp
        }
      }
    }
  }
}

sink {
  ObsFile {
    path="/seatunnel/excel"
    bucket = "obs://******"
    access_key = "******"
    access_secret = "******"
    endpoint = "obs.cn-east-3.myhuaweicloud.com"
    partition_dir_expression="${k0}=${v0}"
    is_partition_field_write_in_file=true
    file_name_expression="${transactionId}_${now}"
    file_format_type="excel"
    filename_time_format="yyyy.MM.dd"
    is_enable_transaction=true
  }
}
```
##### Result
image

image


### obs_excel_projection_to_assert
```
env {
 execution.parallelism = 1
  checkpoint.interval = 5000
  job.mode = "BATCH"
}

source {
  ObsFile {
    path="/seatunnel/excel"
    bucket = "obs://******"
    access_key = "******"
    access_secret = "******"
    endpoint = "obs.cn-east-3.myhuaweicloud.com"
    result_table_name = "fake"
    file_format_type = excel
    delimiter = ;
    read_columns = [c_string, c_boolean]
    skip_header_row_number = 1
    schema = {
      fields {
          c_map = "map<string, string>"
          c_array = "array<int>"
          c_string = string
          c_boolean = boolean
          c_tinyint = tinyint
          c_smallint = smallint
          c_int = int
          c_bigint = bigint
          c_float = float
          c_double = double
          c_bytes = bytes
          c_date = date
          c_decimal = "decimal(38, 18)"
          c_timestamp = timestamp
          c_row = {
            c_map = "map<string, string>"
            c_array = "array<int>"
            c_string = string
            c_boolean = boolean
            c_tinyint = tinyint
            c_smallint = smallint
            c_int = int
            c_bigint = bigint
            c_float = float
            c_double = double
            c_bytes = bytes
            c_date = date
            c_decimal = "decimal(38, 18)"
            c_timestamp = timestamp
          }
      }
    }
  }
}


sink {
  Assert {
    rules {
      row_rules = [
        {
          rule_type = MAX_ROW
          rule_value = 5
        }
      ],
      field_rules = [
        {
          field_name = c_string
          field_type = string
          field_value = [
            {
              rule_type = NOT_NULL
            }
          ]
        },
        {
          field_name = c_boolean
          field_type = boolean
          field_value = [
            {
              rule_type = NOT_NULL
            }
          ]
        }
      ]
    }
  }
}
```
##### Result
image

**Comment by kim-up** [comment]
## Local Test Images
### Json Test
#### fake_to_obs_file_json
```
env {
  execution.parallelism = 1
  checkpoint.interval = 5000
  job.mode = "BATCH"
}

source {
  FakeSource {
    schema = {
      fields {
        c_map = "map<string, string>"
        c_array = "array<int>"
        c_string = string
        c_boolean = boolean
        c_tinyint = tinyint
        c_smallint = smallint
        c_int = int
        c_bigint = bigint
        c_float = float
        c_double = double
        c_bytes = bytes
        c_date = date
        c_decimal = "decimal(38, 18)"
        c_timestamp = timestamp
        c_row = {
          c_map = "map<string, string>"
          c_array = "array<int>"
          c_string = string
          c_boolean = boolean
          c_tinyint = tinyint
          c_smallint = smallint
          c_int = int
          c_bigint = bigint
          c_float = float
          c_double = double
          c_bytes = bytes
          c_date = date
          c_decimal = "decimal(38, 18)"
          c_timestamp = timestamp
        }
      }
    }
    result_table_name = "fake"
  }
}

sink {
  ObsFile {
    path = "/seatunnel/json"
    bucket = "obs://******"
    access_key = "******"
    access_secret = "******"
    endpoint = "obs.cn-east-3.myhuaweicloud.com"
    row_delimiter = "\n"
    partition_dir_expression = "${k0}=${v0}"
    is_partition_field_write_in_file = true
    file_name_expression = "${transactionId}_${now}"
    file_format_type = "json"
    filename_time_format = "yyyy.MM.dd"
    is_enable_transaction = true
  }
}
```
##### Result
image
image



### obs_file_json_to_assert
```
env {
  execution.parallelism = 1
  checkpoint.interval = 5000
  job.mode = "BATCH"
}

source {
  ObsFile {
    path = "/seatunnel/json"
    bucket = "obs://******"
    access_key = "******"
    access_secret = "******"
    endpoint = "obs.cn-east-3.myhuaweicloud.com"
    file_format_type = "json"
    schema = {
      fields {
        c_map = "map<string, string>"
        c_array = "array<int>"
        c_string = string
        c_boolean = boolean
        c_tinyint = tinyint
        c_smallint = smallint
        c_int = int
        c_bigint = bigint
        c_float = float
        c_double = double
        c_bytes = bytes
        c_date = date
        c_decimal = "decimal(38, 18)"
        c_timestamp = timestamp
        c_row = {
          c_map = "map<string, string>"
          c_array = "array<int>"
          c_string = string
          c_boolean = boolean
          c_tinyint = tinyint
          c_smallint = smallint
          c_int = int
          c_bigint = bigint
          c_float = float
          c_double = double
          c_bytes = bytes
          c_date = date
          c_decimal = "decimal(38, 18)"
          c_timestamp = timestamp
        }
      }
    }
    result_table_name = "fake"
  }
}

sink {
  Assert {
    rules {
      row_rules = [
        {
          rule_type = MAX_ROW
          rule_value = 5
        }
      ],
      field_rules = [
        {
          field_name = c_string
          field_type = string
          field_value = [
            {
              rule_type = NOT_NULL
            }
          ]
        },
        {
          field_name = c_boolean
          field_type = boolean
          field_value = [
            {
              rule_type = NOT_NULL
            }
          ]
        },
        {
          field_name = c_double
          field_type = double
          field_value = [
            {
              rule_type = NOT_NULL
            }
          ]
        }
      ]
    }
  }
}
```
##### Result
image

**Comment by kim-up** [comment]
## Local Test Images
### Orc Test
#### fake_to_obs_file_orc
```
env {
  execution.parallelism = 1
  checkpoint.interval = 5000
  job.mode = "BATCH"
}

source {
  FakeSource {
    schema = {
      fields {
        c_map = "map<string, string>"
        c_array = "array<int>"
        c_string = string
        c_boolean = boolean
        c_tinyint = tinyint
        c_smallint = smallint
        c_int = int
        c_bigint = bigint
        c_float = float
        c_double = double
        c_bytes = bytes
        c_date = date
        c_decimal = "decimal(38, 18)"
        c_timestamp = timestamp
        c_row = {
          c_map = "map<string, string>"
          c_array = "array<int>"
          c_string = string
          c_boolean = boolean
          c_tinyint = tinyint
          c_smallint = smallint
          c_int = int
          c_bigint = bigint
          c_float = float
          c_double = double
          c_bytes = bytes
          c_date = date
          c_decimal = "decimal(38, 18)"
          c_timestamp = timestamp
        }
      }
    }
    result_table_name = "fake"
  }
}

sink {
  ObsFile {
    path = "/seatunnel/orc"
    bucket = "obs://*******"
    access_key = "*******"
    access_secret = "*******"
    endpoint = "obs.cn-east-3.myhuaweicloud.com"
    row_delimiter = "\n"
    partition_dir_expression = "${k0}=${v0}"
    is_partition_field_write_in_file = true
    file_name_expression = "${transactionId}_${now}"
    file_format_type = "orc"
    filename_time_format = "yyyy.MM.dd"
    is_enable_transaction = true
    compress_codec = "zlib"
  }
}
```
##### Result
image
image


### obs_file_orc_projection_to_assert
```
env {
  execution.parallelism = 1
  checkpoint.interval = 5000
  job.mode = "BATCH"
}

source {
  ObsFile {
    path = "/seatunnel/orc"
    bucket = "obs://******"
    access_key = "******"
    access_secret = "******"
    endpoint = "obs.cn-east-3.myhuaweicloud.com"
    file_format_type = "orc"
    read_columns = [c_string, c_boolean, c_double]
    result_table_name = "fake"
  }
}

sink {
  Assert {
    rules {
      row_rules = [
        {
          rule_type = MAX_ROW
          rule_value = 5
        }
      ],
      field_rules = [
        {
          field_name = c_string
          field_type = string
          field_value = [
            {
              rule_type = NOT_NULL
            }
          ]
        },
        {
          field_name = c_boolean
          field_type = boolean
          field_value = [
            {
              rule_type = NOT_NULL
            }
          ]
        },
        {
          field_name = c_double
          field_type = double
          field_value = [
            {
              rule_type = NOT_NULL
            }
          ]
        }
      ]
    }
  }
}
```
##### Result
image

**Comment by kim-up** [comment]
## Local Test Images
### Parquet Test
#### fake_to_obs_file_parquet
```
env {
  execution.parallelism = 1
  checkpoint.interval = 5000
  job.mode = "BATCH"
}

source {
  FakeSource {
    schema = {
      fields {
        c_map = "map<string, string>"
        c_array = "array<int>"
        c_string = string
        c_boolean = boolean
        c_tinyint = tinyint
        c_smallint = smallint
        c_int = int
        c_bigint = bigint
        c_float = float
        c_double = double
        c_bytes = bytes
        c_date = date
        c_decimal = "decimal(38, 18)"
        c_timestamp = timestamp
        c_row = {
          c_map = "map<string, string>"
          c_array = "array<int>"
          c_string = string
          c_boolean = boolean
          c_tinyint = tinyint
          c_smallint = smallint
          c_int = int
          c_bigint = bigint
          c_float = float
          c_double = double
          c_bytes = bytes
          c_date = date
          c_decimal = "decimal(38, 18)"
          c_timestamp = timestamp
        }
      }
    }
    result_table_name = "fake"
  }
}

sink {
  ObsFile {
    path = "/seatunnel/parquet"
    bucket = "obs://******"
    access_key = "******"
    access_secret = "******"
    endpoint = "obs.cn-east-3.myhuaweicloud.com"
    row_delimiter = "\n"
    partition_dir_expression = "${k0}=${v0}"
    is_partition_field_write_in_file = true
    file_name_expression = "${transactionId}_${now}"
    file_format_type = "parquet"
    filename_time_format = "yyyy.MM.dd"
    is_enable_transaction = true
    compress_codec = "gzip"
  }
}
```
##### Result
image

image

### obs_file_parquet_projection_to_assert
```
env {
  execution.parallelism = 1
  checkpoint.interval = 5000
  job.mode = "BATCH"
}

source {
  ObsFile {
    path = "/seatunnel/parquet"
    bucket = "obs://******"
    access_key = "******"
    access_secret = "******"
    endpoint = "obs.cn-east-3.myhuaweicloud.com"
    file_format_type = "parquet"
    read_columns = [c_string, c_boolean, c_double]
    result_table_name = "fake"
  }
}

sink {
  Assert {
    rules {
      row_rules = [
        {
          rule_type = MAX_ROW
          rule_value = 5
        }
      ],
      field_rules = [
        {
          field_name = c_string
          field_type = string
          field_value = [
            {
              rule_type = NOT_NULL
            }
          ]
        },
        {
          field_name = c_boolean
          field_type = boolean
          field_value = [
            {
              rule_type = NOT_NULL
            }
          ]
        },
        {
          field_name = c_double
          field_type = double
          field_value = [
            {
              rule_type = NOT_NULL
            }
          ]
        }
      ]
    }
  }
}
```
##### Result
image

**Comment by kim-up** [comment]
## Local Test Images
### Text Test
#### fake_to_obs_file_text
```
env {
  execution.parallelism = 1
  checkpoint.interval = 5000
  job.mode = "BATCH"
}

source {
  FakeSource {
    schema = {
      fields {
        c_map = "map<string, string>"
        c_array = "array<int>"
        c_string = string
        c_boolean = boolean
        c_tinyint = tinyint
        c_smallint = smallint
        c_int = int
        c_bigint = bigint
        c_float = float
        c_double = double
        c_bytes = bytes
        c_date = date
        c_decimal = "decimal(38, 18)"
        c_timestamp = timestamp
        c_row = {
          c_map = "map<string, string>"
          c_array = "array<int>"
          c_string = string
          c_boolean = boolean
          c_tinyint = tinyint
          c_smallint = smallint
          c_int = int
          c_bigint = bigint
          c_float = float
          c_double = double
          c_bytes = bytes
          c_date = date
          c_decimal = "decimal(38, 18)"
          c_timestamp = timestamp
        }
      }
    }
    result_table_name = "fake"
  }
}

sink {
  ObsFile {
    path = "/seatunnel/text"
    bucket = "obs://******"
    access_key = "******"
    access_secret = "******"
    endpoint = "obs.cn-east-3.myhuaweicloud.com"
    row_delimiter = "\n"
    partition_dir_expression = "${k0}=${v0}"
    is_partition_field_write_in_file = true
    file_name_expression = "${transactionId}_${now}"
    file_format_type = "text"
    filename_time_format = "yyyy.MM.dd"
    is_enable_transaction = true
  }
}
```
##### Result
image

image


### obs_file_text_projection_to_assert
```
env {
  execution.parallelism = 1
  checkpoint.interval = 5000
  job.mode = "BATCH"
}

source {
  ObsFile {
    path = "/seatunnel/text"
    bucket = "obs://******"
    access_key = "******"
    access_secret = "******"
    endpoint = "obs.cn-east-3.myhuaweicloud.com"
    file_format_type = "text"
    read_columns = [c_string, c_boolean, c_double]
    schema = {
      fields {
        c_map = "map<string, string>"
        c_array = "array<int>"
        c_string = string
        c_boolean = boolean
        c_tinyint = tinyint
        c_smallint = smallint
        c_int = int
        c_bigint = bigint
        c_float = float
        c_double = double
        c_bytes = bytes
        c_date = date
        c_decimal = "decimal(38, 18)"
        c_timestamp = timestamp
        c_row = {
          c_map = "map<string, string>"
          c_array = "array<int>"
          c_string = string
          c_boolean = boolean
          c_tinyint = tinyint
          c_smallint = smallint
          c_int = int
          c_bigint = bigint
          c_float = float
          c_double = double
          c_bytes = bytes
          c_date = date
          c_decimal = "decimal(38, 18)"
          c_timestamp = timestamp
        }
      }
    }
    result_table_name = "fake"
  }
}

sink {
  Assert {
    rules {
      row_rules = [
        {
          rule_type = MAX_ROW
          rule_value = 5
        }
      ],
      field_rules = [
        {
          field_name = c_string
          field_type = string
          field_value = [
            {
              rule_type = NOT_NULL
            }
          ]
        },
        {
          field_name = c_boolean
          field_type = boolean
          field_value = [
            {
              rule_type = NOT_NULL
            }
          ]
        },
        {
          field_name = c_double
          field_type = double
          field_value = [
            {
              rule_type = NOT_NULL
            }
          ]
        }
      ]
    }
  }
}
```
##### Result

image

**Comment by kim-up** [comment]
@TyrantLucifer Did you mean the test screenshot provided above?

**Comment by ic4y** [comment]
@kim-up  Please use mvn spotless:apply to format the code.

**Comment by kim-up** [comment]
> @kim-up Please use mvn spotless:apply to format the code.

Done. Fixed  by merging `dev` .

image

**Comment by EricJoy2048** [comment]
Hi, @kim-up 
You submitted a great PR, but it hasn't been updated in a while, can you update it?

**Review by EricJoy2048 — APPROVED** [review]

**Review by hailin0 — APPROVED** [review]

**Pull request merged**